index.bs

```diff
@@ -82,7 +96,37 @@ media frames as input.
 
 <!-- ## Extension operation ## {#operation} -->
 
-## MediaStreamTrackProcessor interface ## {#track-processor}
+## MediaStreamTrackProcessor ## {#track-processor}
+
+A {{MediaStreamTrackProcessor}} allows the creation of a
+{{ReadableStream}} that can expose the media flowing through
+a given {{MediaStreamTrack}}. If the {{MediaStreamTrack}} is a video track,
+the chunks exposed by the stream will be {{VideoFrame}} objects;
+if the track is an audio track, the chunks will be {{AudioData}} objects.
+This makes {{MediaStreamTrackProcessor}} effectively a sink in the
+<a href="https://www.w3.org/TR/mediacapture-streams/#the-model-sources-sinks-constraints-and-settings">
+MediaStream model</a>.
+
+A {{MediaStreamTrackProcessor}} internally contains a circular queue
+that allows buffering incoming media frames delivered by the track it
+is connected to. This buffering allows the {{MediaStreamTrackProcessor}}
+to temporarily hold frames waiting to be read from its associated {{ReadableStream}}.
+If the application is reading frames too slowly and the queue has reached its
+maximum size when the track delivers a new frame, the {{MediaStreamTrackProcessor}}
+will drop the oldest frame from the queue and push the new frame into the queue.
+The maximum size of the queue is decided by the UA and can change dynamically.
+The application can influence the maximum size of the queue via a parameter provided
+in the {{MediaStreamTrackProcessor}} constructor. In this case, the
+{{MediaStreamTrackProcessor}} will make sure that the size of the queue
+is never greater than the value provided by the application, but the actual maximum
+size of the queue can be smaller than such value.
```

Comment:
This seems new. Do we need to update the WPT tests to allow the tests to pass when the UA reduces the queue size (possibly to 0)? @tonyherre

Reply:
We didn't land the WPT which explicitly tested the maxBufferSize because it was failing due to the buffered frames not being dropped when tight looping - which now I think of it sounds exactly like the bug Chrome encountered and was fixed by Guido.
I'll look at re-adding the test and will bear this point in mind - thanks for pointing it out!

Reply:
Yes, this is new. I added this to allow for the implementation to 

> This seems new. Do we need to update the WPT tests to allow the tests to pass when the UA reduces the queue size (possibly to 0)? @tonyherre

Yes, this is new. I added it as a way to allow the UA to more easily implement mitigations to address the issue of pooled frames. 
The bad things is that it becomes more difficult to have deterministic tests. Maybe we can expose the true maxBufferSize as a field, but not sure yet. WDYT?

Reply:
I think it's fine for testing as is, and is preferable for keeping implementation details internal.

We can have a test which sets maxBufferSize to say 10, feeds 20 frames with timestamps 1..20, starts a read and then feeds another frame.
Asserting that the read resolves, and we don't see a frame with timestamp less than 10 should work. If the internal maxBufferSize is less than 10, we'll just get a frame starting with some timestamp >10 (or in the extreme case of internal buffer size being 0, we'll just see the 21st frame).

Reply:
Here are my thoughts:
 - This API is meant for real-time processing. If processing frames consistently takes longer than the frame interval (1/FPS), then we have to drop frames.
 - For almost all use cases, the time it takes to process each frame is fairly consistent. In that case, if the processing time is longer than the frame interval, we would only want to buffer one frame.
 - There are some use cases where we want to do a deep analysis every N frames, then process other frames quickly. I think this is the only use case where maxBufferSize would be > 1. (OTOH, maybe this would be better served by doing a deep analysis asynchronously on a clone of the original track where maxBufferSize is 1.) However, you would probably not want LIFO queuing in that case; instead, you would want to drop frames such that the interval between frames remaining in the queue is fairly consistent. I.e. if maxBufferSize is 3 and 6 new frames come in, I would want the queue to be something like [2, 4, 6] rather than [4, 5, 6]. In the case of audio frames, ideally the data would be merged to create larger chunks.

I don't know what would be best, but here are my suggestions:
A. MediaStreamTrackProcessorInit just has a boolean property buffered; if buffered is true, maxBufferSize is 1; if false, it's 0.
B. maxBufferSize must be 0 or 1. (Same as A but allows extension.)
C. Keep everything as written in this PR, but add a synchronous API to retrieve the timestamps of all buffered frames. (This allows the app to decide if it wants to ignore the frames that are first in the queue.)
D. Keep everything as written in this PR, except require that UAs decimate frames to spread the buffered frames timestamps over the interval since the timestamp of the last frame that was read. Also require that AudioDatas are merged.

I don't really like the idea of exposing maxBufferSize. Since the UA can change it at any time, it becomes difficult to react appropriately, and it doesn't actually indicate whether any frames can currently be read without blocking.

One other variation: Add a boolean property queueOverflowed which is set to true anytime a frame is dropped and false each time a frame is read from the queue. This could potentially allow the app to adjust the source FPS to ensure no frames are dropped.

Reply:
As discussed offline, updated the text so that if no maxBufferSize is provided via the constructor, the UA is free to apply any policy.
If provided, the dropping behavior is as already specified, but the UA can still use a smaller buffer size.